kroxylicious-test-tools/src/main/java/io/kroxylicious/test/client/KafkaClientHandler.java

```diff
@@ -18,26 +20,57 @@
  * Sends a single request frame, waits for a response then closes the channel
  */
 public class KafkaClientHandler extends ChannelInboundHandlerAdapter {
-    private final DecodedRequestFrame<?> decodedRequestFrame;
-    private final CompletableFuture<DecodedResponseFrame<?>> onResponse = new CompletableFuture<>();
+    private final List<DecodedRequestFrame<?>> decodedRequestFrames;
+    private final DecodedRequestFrame<?> initialRequest;
+    List<DecodedResponseFrame<?>> responses = new ArrayList<>();
+    private final CompletableFuture<List<DecodedResponseFrame<?>>> onResponse = new CompletableFuture<>();
+    private final CompletableFuture<Void> initialResponseReceived = new CompletableFuture<>();
 
     /**
      * Creates a KafkaClientHandler
-     * @param decodedRequestFrame the single request to send
+     * @param decodedRequestFrames the requests to send
      */
-    public KafkaClientHandler(DecodedRequestFrame<?> decodedRequestFrame) {
-        this.decodedRequestFrame = decodedRequestFrame;
+    public KafkaClientHandler(DecodedRequestFrame<?> initialRequest, List<DecodedRequestFrame<?>> decodedRequestFrames) {
+        this.decodedRequestFrames = decodedRequestFrames;
+        this.initialRequest = initialRequest;
+        if (this.initialRequest == null) {
+            initialResponseReceived.complete(null);
+        }
     }
 
     @Override
     public void channelActive(ChannelHandlerContext ctx) {
-        ctx.writeAndFlush(decodedRequestFrame);
+        if (initialRequest != null) {
+            ctx.writeAndFlush(initialRequest);
+        }
+        initialResponseReceived.thenAccept(unused -> {
+            writeAllMessages(ctx);
+        });
+    }
+
+    private void writeAllMessages(ChannelHandlerContext ctx) {
+        for (DecodedRequestFrame<?> decodedRequestFrame : decodedRequestFrames) {
+            ctx.write(decodedRequestFrame);
+        }
+        ctx.flush();
     }
 
     @Override
     public void channelRead(ChannelHandlerContext ctx, Object msg) {
-        onResponse.complete((DecodedResponseFrame<?>) msg);
-        ctx.write(msg).addListener(ChannelFutureListener.CLOSE);
+        if (!initialResponseReceived.isDone()) {
+            initialResponseReceived.complete(null);
+        }
+        else {
+            responses.add((DecodedResponseFrame<?>) msg);
+            if (allResponsesReceived()) {
+                onResponse.complete(responses);
+                ctx.write(msg).addListener(ChannelFutureListener.CLOSE);
```

Comment:
why do we write the last message?

Reply:
hmm, no good reason from what I can tell, have changed to `ctx.close()`